1161/identifierforvendor andinstance (#1412)
OWASP/owasp-mstg: Document/0x04e-Testing-Authentication-and-Session-Management.md

3. The application provides an overview of the last session after login at all times.
4. The application has a self-service portal in which the user can see an audit-log and manage the different devices with which he can login.

In all cases, you should verify whether different devices are detected correctly. Therefore, the binding of the application to the actual device should be tested. 

> In iOS, a developer can use `identifierForVendor`, which is related to the bundleID: the moment you change a bundle-ID, the method will return a different value. When the app is ran for the first time, make sure you store the String returned by `identifierForVendor` to the KeyChain, so that changes to it can be detected in an early stage.

Reviewer: ```suggestion
In iOS, a developer can use `identifierForVendor`, which is related to the bundleID: the moment you change a bundle-ID, the method will return a different value. When the app is ran for the first time, make sure you store the String returned by `identifierForVendor` to the KeyChain, so that changes to it can be detected at an early stage.
```

in -> at

> In Android, the developer can use `Settings.Secure.ANDROID_ID` till Android 8.0 (API level 26) to identify an application instance. Note that starting at Android 8.0 (API level 26), `ANDROID_ID` is no longer a device unique ID. Instead it becomes scoped by the combination of app-signing key, user and device. So validating `ANDROID_ID` for device blocking could be tricky for these Android versions. Because if an app changes its signing key, the `ANDROID_ID` will change and it won't be able to recognize old users devices. Therefore it is better to store the `ANDROID_ID` encrypted with a randomly generated key from the `AndroidKeyStore` using AES_GCM encryption in the `SharedPreferences` (privately). The moment the app-signature changes, the application can check for a delta and reegister the new `ANDROID_ID`. The moment this changes without a new application signing key, it should indicate that something else is wrong.

Reviewer: ```suggestion
In Android, the developer can use `Settings.Secure.ANDROID_ID` till Android 8.0 (API level 26) to identify an application instance. Note that starting at Android 8.0 (API level 26), `ANDROID_ID` is no longer a device unique ID. Instead it becomes scoped by the combination of app-signing key, user and device. So validating `ANDROID_ID` for device blocking could be tricky for these Android versions. Because if an app changes its signing key, the `ANDROID_ID` will change and it won't be able to recognize old users devices. Therefore it is better to store the `ANDROID_ID` encrypted with a randomly generated key from the `AndroidKeyStore` using AES_GCM encryption in the `SharedPreferences` (privately). The moment the app-signature changes, the application can check for a delta and register the new `ANDROID_ID`. The moment this changes without a new application signing key, it should indicate that something else is wrong.
```
reegister -> register

> In Android, the developer can use `Settings.Secure.ANDROID_ID` till Android 8.0 (API level 26) to identify an application instance. Note that starting at Android 8.0 (API level 26), `ANDROID_ID` is no longer a device unique ID. Instead it becomes scoped by the combination of app-signing key, user and device. So validating `ANDROID_ID` for device blocking could be tricky for these Android versions. Because if an app changes its signing key, the `ANDROID_ID` will change and it won't be able to recognize old users devices. Therefore it is better to store the `ANDROID_ID` encrypted with a randomly generated key from the `AndroidKeyStore` using AES_GCM encryption in the `SharedPreferences` (privately). The moment the app-signature changes, the application can check for a delta and reegister the new `ANDROID_ID`. The moment this changes without a new application signing key, it should indicate that something else is wrong.

Reviewer 2: Minor re-phrase and referring to "shared preferences file" instead of "store...in the `SharedPreferences` (class).
```suggestion
In Android, the developer can use `Settings.Secure.ANDROID_ID` till Android 8.0 (API level 26) to identify an application instance. Note that starting at Android 8.0 (API level 26), `ANDROID_ID` is no longer a device unique ID. Instead, it becomes scoped by the combination of app signing key, user and device. So validating `ANDROID_ID` for device blocking could be tricky for these Android versions. Because if an app changes its signing key, the `ANDROID_ID` will change and it won't be able to recognize old users devices. Therefore, it's better to store the `ANDROID_ID` encrypted and privately in a private a shared preferences file using a randomly generated key from the `AndroidKeyStore` and preferably AES_GCM encryption. The moment the app signature changes, the application can check for a delta and register the new `ANDROID_ID`. The moment this new ID changes without a new application signing key, it should indicate that something else is wrong.
```

> Next the device binding can be extended by signing requests wih a key stored in the `Keychain` for iOS and in the `KeyStore` in Android can reassure strong device binding. Next, you should test if using different IPs, different locations and/or different time-slots will trigger the right type of information in all scenarios.

Reviewer: ```suggestion
Next the device binding can be extended by signing requests with a key stored in the `Keychain` for iOS and in the `KeyStore` in Android can reassure strong device binding. Next, you should test if using different IPs, different locations and/or different time-slots will trigger the right type of information in all scenarios.
```

wih -> with

> Next the device binding can be extended by signing requests wih a key stored in the `Keychain` for iOS and in the `KeyStore` in Android can reassure strong device binding. Next, you should test if using different IPs, different locations and/or different time-slots will trigger the right type of information in all scenarios.

Reviewer 2: 2 "next" too close to each other.

```suggestion
Next, the device binding can be extended by signing requests with a key stored in the `Keychain` for iOS and in the `KeyStore` in Android can reassure strong device binding. You should also test if using different IPs, different locations and/or different time-slots will trigger the right type of information in all scenarios.
```

3. The application provides an overview of the last session after login at all times.
4. The application has a self-service portal in which the user can see an audit-log and manage the different devices with which he can login.

In all cases, you should verify whether different devices are detected correctly. Therefore, the binding of the application to the actual device should be tested. 

> In iOS, a developer can use `identifierForVendor`, which is related to the bundleID: the moment you change a bundle-ID, the method will return a different value. When the app is ran for the first time, make sure you store the String returned by `identifierForVendor` to the KeyChain, so that changes to it can be detected at an early stage.

Reviewer 2: ```suggestion
In iOS, a developer can use `identifierForVendor`, which is related to the bundle ID: the moment you change a bundle ID, the method will return a different value. When the app is ran for the first time, make sure you store the string returned by `identifierForVendor` to the KeyChain, so that changes to it can be detected at an early stage.
```

Author: Resolved a bit differently

> In Android, the developer can use `Settings.Secure.ANDROID_ID` till Android 8.0 (API level 26) to identify an application instance. Note that starting at Android 8.0 (API level 26), `ANDROID_ID` is no longer a device unique ID. Instead it becomes scoped by the combination of app-signing key, user and device. So validating `ANDROID_ID` for device blocking could be tricky for these Android versions. Because if an app changes its signing key, the `ANDROID_ID` will change and it won't be able to recognize old users devices. Therefore it is better to encrypt the `ANDROID_ID` with a randomly generated key from the `AndroidKeyStore` using `AES_GCM` encryption. The encrypted `ANDROID_ID` should then be stored in the `SharedPreferences` (privately). The moment the app-signature changes, the application can check for a delta and register the new `ANDROID_ID`. The moment this changes without a new application signing key, it should indicate that something else is wrong.

Reviewer 2: ```suggestion
In Android, the developer can use `Settings.Secure.ANDROID_ID` till Android 8.0 (API level 26) to identify an application instance. Note that starting at Android 8.0 (API level 26), `ANDROID_ID` is no longer a device unique ID. Instead, it becomes scoped by the combination of app signing key, user and device. So validating `ANDROID_ID` for device blocking could be tricky for these Android versions. Because if an app changes its signing key, the `ANDROID_ID` will change and it won't be able to recognize old users devices. Therefore, it's better to store the `ANDROID_ID` encrypted and privately in a private a shared preferences file using a randomly generated key from the `AndroidKeyStore` and preferably AES_GCM encryption. The moment the app signature changes, the application can check for a delta and register the new `ANDROID_ID`. The moment this new ID changes without a new application signing key, it should indicate that something else is wrong.
```
